Classes/CharacterManager.gd

```diff
@@ -0,0 +1,79 @@
+class_name CharacterManager
+const script_name := "character_manager"
+
+static func create(character_name: String):
+	return CharacterDefaults.new().characters[character_name]
+
+
+static func load_class(character_data: Dictionary, character_class: String):
+	# Add equipment to the character inventory and equip those items
+	for item in CharacterDefaults.starting_equipment[character_class]:
+		var item_index = add(character_data, Items.items[item.name], item.quantity)
+		equip(character_data, item_index)
+	
+	# Add character attacks
+	for attack_type in CharacterDefaults.starting_attacks[character_class].keys():
+		for attack_name in CharacterDefaults.starting_attacks[character_class][attack_type]:
+			learn_attack(character_data, attack_type, attack_name)
+	
+	return character_data
+
+
+static func learn_attack(character_data: Dictionary, attack_type: String, attack_name: String):
+	character_data.attacks[attack_type].append(attack_name)
+
+
+static func calcuate_item_buffs(character_data):
+	Core.emit_signal("msg", "This function is not implemented yet!", Log.WARN, "character_manager")
+	#for item in player.items:
+	#	pass
+
+
+static func add(character_data: Dictionary, item_data: Dictionary, quantity:=1):
+	return InventoryManager.add(character_data.items.inventory, item_data, quantity)
+
+
+static func remove(character_data: Dictionary, item_data: Dictionary, quantity:=1):
+	InventoryManager.remove(character_data.items.inventory, item_data, quantity)
+
+
+static func equip(character_data: Dictionary, item_index: int):
+	if !character_data.items.inventory[item_index]:
+		return false
+	var item = character_data.items.inventory[item_index]
+
+	if item.level_requirement <= character_data.level:
+		character_data.items[item.type][item.subtype].append(item_index)
+	else:
+		return false
+
+
+static func unequip(character_data: Dictionary, item_index: int):
+	if !character_data.items.inventory[item_index]:
+		return false
+	var item = character_data.items.inventory[item_index]
+	
+	character_data.items[item.type][item.subtype].erase(item_index)
```

Comment:
Inventory was not meant to be a sub value of items
Inventory and items are both inventory classes, which don't have sub inventories if they did this would be an infinite inventory of inventories

Reply:
hmmm looks good to me?
```
{
  "meta": {
    "name": "Alrune",
    "nickname": "",
    "info": ""
  },
  "file": "",
  "stats": [
    3,
    3,
    3,
    3,
    3,
    3,
    3
  ],
  "picture": {
    "path": "Friendlies/Tex_AnimeAva_28.png",
    "border": {
      "path": "ImageBorder.png",
      "shown": true
    }
  },
  "attacks": {
    "melee": [
      "punch",
      "strike"
    ],
    "ranged": [],
    "mana": []
  },
  "level": 1,
  "skills": {},
  "ap": {
    "max": 1,
    "current": 0.5
  },
  "health": {
    "max": 100,
    "current": 100
  },
  "mana": {
    "max": 50,
    "current": 100
  },
  "kills": 0,
  "items": {
    "armour": {
      "head": [],
      "torso": [],
      "arms": [],
      "legs": [],
      "feet": [],
      "shield": []
    },
    "equipment": {
      "melee": [
        0
      ],
      "ranged": [],
      "consumables": [],
      "magic": []
    },
    "inventory": [
      {
        "nickname": "Sword",
        "name": "sword",
        "type": "equipment",
        "subtype": "melee",
        "hand": "one-handed",
        "tags": [],
        "buffs": {
          "strength": 2,
          "perception": 0,
          "endurance": 0,
          "charisma": 0,
          "intelligence": 0,
          "agility": 0,
          "luck": 0,
          "melee": 10,
          "mana": 0,
          "defense": 0
        },
        "level": 1,
        "level_requirement": 1,
        "quantity": 1
      }
    ]
  },
  "nickname": "fpg"
}
```

Reply:
items shouldn't exist
equipment and inventory are the only ones
you've mixed up items with what equipment is for
equipment is the currently equipped stuff and inventory is everything the character owns
items is not needed
and if you do want to keep it the armour will be in equipment and items will just have the two inventory classes

Reply:
also, why is 0 equiped as a weapon?
0 is not a weapon
equipment should hold an exact copy of the item otherwise the equip/unequip and battle code needs to be changed

Reply:
I thought you had set this back, but now it seems like you've gone back to what you changed it to